src/toc.ts

```diff
@@ -0,0 +1,43 @@
+import type { VFileCompatible } from 'vfile';
+
+export type TocItem = {
+    depth: number;
+    text: string;
+    id: string;
+    children: TocItem[];
+};
+
+function generateTableOfContents(markdown: VFileCompatible): TocItem[] {
+    const toc: TocItem[] = [];
+    const lines = markdown.toString().split('\n').filter(line => line.trim() !== '').map(line => line.trim());
+    const heading_regex = /^#{1,6}\s+(.+)$/
+    const heading_stack: TocItem[] = [];
+    for (const line of lines) {
+        const heading_match = line.match(heading_regex);
+        if (heading_match) {
+            const heading = heading_match[1];
+            if (heading_match[0].match(/^#+/) === null) {
+                throw new Error(`Invalid heading: ${heading_match[0]} \n on line: ${line}`);
+            }
+            // @ts-ignore - matching does see the if statement above
+            const depth = heading_match[0].match(/^#+/)[0].length;
+            const id = heading.toLowerCase().replace(/[^a-z0-9]+/g, '-');
```

Comment:
In case of duplicated HTML ids, maybe an `idMap: Map<string, number>` is needed to change a new duplicated id to `${id}-${idMap[id].length}`

Reply:
why not I will test this idea